Disk margin calculations

josiahdelange opened the pull request:
Am relatively new to this toolbox, have been using it here and there and found #726.

This is an initial prototype Python implementation of disk margins, built on top of `python-control` and `slycot`.  The code should work both for SISO and MIMO systems, the latter of which requires `slycot` (for SLICOT's `AB13MD`) to compute the upper bound of $`\mu`$ at each discrete frequency.  The function `disk_margins` computes disk margins (and corresponding disk-based gain/phase margins), optionally returning the whole frequency-dependent vectors for further plotting.

It's been verified against the example SISO loop transfer functions in the [published paper](https://arxiv.org/pdf/2003.04771) and the "spinning satellite" MIMO [example](https://www.mathworks.com/help/robust/ug/mimo-stability-margins-for-spinning-satellite.html) from the MathWorks documentation.  I also confirmed SISO disk margins match the relevant output of existing function [`stability_margins`](https://python-control.readthedocs.io/en/latest/generated/control.stability_margins.html) corresponding to the Nyquist plot's distance to -1.  All seem to match within a few significant digits, so the general behavior seems correct.  Might be good to get another set of eyes to double check/test further.

I tried to base as much as possible (e.g. style conventions) on existing code in `control/margins.py`.  Example usage (see `examples/disk_margin.py`):

```python
import os, sys, math
import numpy as np
import control

import math
import matplotlib
import matplotlib.pyplot as plt
from warnings import warn

import numpy as np
import scipy as sp

# Frequencies of interest
omega = np.logspace(-1, 3, 1001)

# Laplace variable
s = control.tf('s')

# MIMO loop transfer gain for the spinning satellite example
# https://www.mathworks.com/help/robust/ug/mimo-stability-margins-for-spinning-satellite.html
P = control.ss([[0, 10],[-10, 0]], np.eye(2), [[1, 10], [-10, 1]], [[0, 0],[0, 0]]) # plant
C = control.ss([],[],[], [[1, -2], [0, 1]]) # controller
L = P*C # output loop gain

print(f"------------- Sensitivity function (S) -------------")
DM, GM, PM = control.disk_margins(L, omega, skew = 1.0, returnall = True) # S-based (S)
print(f"min(DM) = {min(DM)} (omega = {omega[np.argmin(DM)]})")
print(f"GM = {GM[np.argmin(DM)]} dB")
print(f"PM = {PM[np.argmin(DM)]} deg")
print(f"min(GM) = {min(GM)} dB")
print(f"min(PM) = {min(PM)} deg\n")

plt.figure(3)
plt.subplot(3,3,1)
plt.semilogx(omega, DM, label='$\\alpha$')
plt.legend()
plt.title('Disk Margin')
plt.grid()
plt.xlim([omega[0], omega[-1]])
plt.ylim([0, 2])

plt.figure(3)
plt.subplot(3,3,4)
plt.semilogx(omega, GM, label='$\\gamma_{m}$')
plt.ylabel('Gain Margin (dB)')
plt.legend()
plt.title('Gain-Only Margin')
plt.grid()
plt.xlim([omega[0], omega[-1]])
plt.ylim([0, 40])

plt.figure(3)
plt.subplot(3,3,7)
plt.semilogx(omega, PM, label='$\\phi_{m}$')
plt.ylabel('Phase Margin (deg)')
plt.legend()
plt.title('Phase-Only Margin')
plt.grid()
plt.xlim([omega[0], omega[-1]])
plt.ylim([0, 90])

print(f"------------- Complementary sensitivity function (T) -------------")
DM, GM, PM = control.disk_margins(L, omega, skew = -1.0, returnall = True) # T-based (T)
print(f"min(DM) = {min(DM)} (omega = {omega[np.argmin(DM)]})")
print(f"GM = {GM[np.argmin(DM)]} dB")
print(f"PM = {PM[np.argmin(DM)]} deg")
print(f"min(GM) = {min(GM)} dB")
print(f"min(PM) = {min(PM)} deg\n")

plt.figure(3)
plt.subplot(3,3,2)
plt.semilogx(omega, DM, label='$\\alpha$')
plt.legend()
plt.title('Disk Margin')
plt.grid()
plt.xlim([omega[0], omega[-1]])
plt.ylim([0, 2])

plt.figure(3)
plt.subplot(3,3,5)
plt.semilogx(omega, GM, label='$\\gamma_{m}$')
plt.ylabel('Gain Margin (dB)')
plt.legend()
plt.title('Gain-Only Margin')
plt.grid()
plt.xlim([omega[0], omega[-1]])
plt.ylim([0, 40])

plt.figure(3)
plt.subplot(3,3,8)
plt.semilogx(omega, PM, label='$\\phi_{m}$')
plt.ylabel('Phase Margin (deg)')
plt.legend()
plt.title('Phase-Only Margin')
plt.grid()
plt.xlim([omega[0], omega[-1]])
plt.ylim([0, 90])

print(f"------------- Balanced sensitivity function (S - T) -------------")
DM, GM, PM = control.disk_margins(L, omega, skew = 0.0, returnall = True) # balanced (S - T)
print(f"min(DM) = {min(DM)} (omega = {omega[np.argmin(DM)]})")
print(f"GM = {GM[np.argmin(DM)]} dB")
print(f"PM = {PM[np.argmin(DM)]} deg")
print(f"min(GM) = {min(GM)} dB")
print(f"min(PM) = {min(PM)} deg\n")

plt.figure(3)
plt.subplot(3,3,3)
plt.semilogx(omega, DM, label='$\\alpha$')
plt.legend()
plt.title('Disk Margin')
plt.grid()
plt.xlim([omega[0], omega[-1]])
plt.ylim([0, 2])

plt.figure(3)
plt.subplot(3,3,6)
plt.semilogx(omega, GM, label='$\\gamma_{m}$')
plt.ylabel('Gain Margin (dB)')
plt.legend()
plt.title('Gain-Only Margin')
plt.grid()
plt.xlim([omega[0], omega[-1]])
plt.ylim([0, 40])

plt.figure(3)
plt.subplot(3,3,9)
plt.semilogx(omega, PM, label='$\\phi_{m}$')
plt.ylabel('Phase Margin (deg)')
plt.legend()
plt.title('Phase-Only Margin')
plt.grid()
plt.xlim([omega[0], omega[-1]])
plt.ylim([0, 90])

```

Output:
```
------------- Sensitivity function (S) -------------
min(DM) = 0.3697398698410271 (omega = 0.1)
GM = 2.732761944304522 dB
PM = 21.30709883385843 deg
min(GM) = 2.732761944304522 dB
min(PM) = 21.30709883385843 deg

------------- Complementary sensitivity function (T) -------------
min(DM) = 0.3683035923739884 (omega = 0.1)
GM = 2.7236493433653735 dB
PM = 21.223368586091564 deg
min(GM) = 2.7236493433653735 dB
min(PM) = 21.223368586091564 deg

------------- Balanced sensitivity function (S - T) -------------
min(DM) = 0.3769872636944355 (omega = 0.1)
GM = 3.3140985364257256 dB
PM = 21.349285542574695 deg
min(GM) = 3.3140985364257256 dB
min(PM) = 21.349285542574695 deg
```
![Figure_3](https://github.com/user-attachments/assets/0bed7d01-2343-494c-bc77-dc19038e3191)


The example script also shows how to plot the allowable/stable region of gain and phase variations which will not destabilize the loop, in a local function `plot_allowable_region`, e.g.
```
.
.
.
DM_plot = []
DM_plot.append(control.disk_margins(L, omega, skew = -1.0)[0]) # T-based (T)
DM_plot.append(control.disk_margins(L, omega, skew = 0.0)[0]) # balanced (S - T)
DM_plot.append(control.disk_margins(L, omega, skew = 1.0)[0]) # S-based (S)
plt.figure(30)
plot_allowable_region(DM_plot, skew = [-1.0, 0.0, 1.0])
```

![Figure_30](https://github.com/user-attachments/assets/906589ad-e4e2-4397-88db-483d0b549e25)

josiahdelange committed:
Initial version of disk margin calculation and example/test script

josiahdelange committed:
Merge remote-tracking branch 'origin' into jdelange/disk-margins

josiahdelange committed:
Comment updates: update margins.py header, clarify import exception handler comment, fix typo in skew description of disk_margins docstring

josiahdelange committed:
More work in progress on disk margin calculation, adding new prototype function to plot allowable gain/phase variations.

josiahdelange committed:
Add disk_margin_plot to subroutine list in comment header in margins.py

josiahdelange committed:
Follow-on to ba157895fee83ecc15bd5c1bcd8f56f4e50778a5, add disk_margin_plot to list of functions within the margins module

josiahdelange committed:
Merge branch 'python-control:main' into jdelange/disk-margins

josiahdelange committed:
More work in progress on disk_margin_plot.  Corrected a typo/bug in the
calculation of 'f', the bounding complex curve.  Seems to look correct for balanced
(skew = 0) case, still verifying the skewed equivalent.

josiahdelange committed:
Merge branch 'jdelange/disk-margins' of github.com:josiahdelange/python-control into jdelange/disk-margins

josiahdelange committed:
Merge remote-tracking branch 'origin/main' into jdelange/disk-margins

josiahdelange committed:
Further progress/debugging on disk margin calculation + plot utility

josiahdelange committed:
Merge remote-tracking branch 'origin/main' into jdelange/disk-margins

josiahdelange committed:
Clean up docstring/code for disk_margin_plot

josiahdelange committed:
Clean up docstring/code for disk_margin_plot

josiahdelange committed:
Remove debugging statements, update comments, add unit tests.

josiahdelange committed:
Minor change to fix logic to find minimum across DGM, DPM numpy vectors

josiahdelange committed:
Rename disk margin example, since unit tests are now written in control/tests/margin_test.py

josiahdelange committed:
Remove unneeded dependencies from margins.py, used for debugging

josiahdelange committed:
Minor updates to docstrings

josiahdelange committed:
Undo d92fb2045a786581741ddb703819f7ae5865a323

josiahdelange committed:
Minor tweaks to plots in example script for readability

josiahdelange committed:
Fix typo in disk_margin_plot.

slivingston commented on control/margins.py:
```diff
@@ -10,13 +10,29 @@
+    from slycot import ab13md
+except ImportError:
+    ab13md = None
+try:
+    from . import mag2db
+except ImportError:
+    # Likely due the following circular import issue:
+    #
+    # ImportError: cannot import name 'mag2db' from partially initialized module
+    # 'control' (most likely due to a circular import) (control/__init__.py)
+    #
+    def mag2db(mag):
+        return 20*np.log10(mag)
```
Try to import from the module where this is defined:

    from .ctrlutil import mag2db

slivingston commented on control/margins.py:
```diff
@@ -517,3 +532,281 @@ def margin(*args):
+    >>
+    >> omega = np.logspace(-1, 3, 1001)
+    >>
+    >> P = control.ss([[0, 10],[-10, 0]], np.eye(2), [[1, 10], [-10, 1]], [[0, 0],[0, 0]])
```
This example seems too big for a docstring, and anyway, it already is given in examples/disk_margins.py. Is there a smaller example that can be given in this docstring?

josiahdelange replied:
Agree.  Yes - will remove all the plotting.

josiahdelange committed:
Fix mag2db import hack/workaround and trim down disk_margin docstring.

josiahdelange committed:
Add input handling to disk_margin, clean up column width/comments

josiahdelange committed:
Move disk_margin_plot out of the library into the example script

josiahdelange committed:
Recommended changes from the linter

josiahdelange committed:
Follow-on to 5f34a7bea410715ee1389a36cd8de3e7001ebf34

josiahdelange committed:
Add disk_margins to function list

josiahdelange committed:
Whittle down the docstring from disk_margins

josiahdelange committed:
Put more comments in the disk margin example, add example to documentation

josiahdelange committed:
Fixing docstrings

josiahdelange commented:
I fixed the docstrings and implemented the linter's recommendations so that automated workflows finally pass.  One thing I previously hadn't appreciated was the complexity of the control plotting code, so to make this PR simpler I'm only proposing to add a library function to calculate the disk-based margins.  All other plotting is left to user code (or future PRs), e.g. the local function `plot_allowable_region` in `examples/disk_margins.py`.

josiahdelange committed:
Corrected expected values for 'no-slycot' condition in newly-added unit tests

josiahdelange committed:
Attempt #2 at 397efabbe7ff9dcc11f2c4309ba73d63ed44d742, based on linter recommendation

murrayrm commented on control/tests/margin_test.py:
```diff
@@ -372,3 +373,135 @@ def test_stability_margins_discrete(cnum, cden, dt,
+
+            # Balanced (S - T) disk-based stability margins at plant output
+            DMo, DGMo, DPMo = disk_margins(Lo, omega, skew = 0.0)
+            assert_allclose([DMo], [0.3754], atol = 0.1) # disk margin of 0.3754
+            assert_allclose([DGMo], [3.3], atol = 0.1) # disk-based gain margin of 3.3 dB 
+            assert_allclose([DPMo], [21.26], atol = 0.1) # disk-based phase margin of 21.26 deg
+
+            # Balanced (S - T) disk-based stability margins at plant input
+            DMi, DGMi, DPMi = disk_margins(Li, omega, skew = 0.0)
+            assert_allclose([DMi], [0.3754], atol = 0.1) # disk margin of 0.3754
+            assert_allclose([DGMi], [3.3], atol = 0.1) # disk-based gain margin of 3.3 dB 
+            assert_allclose([DPMi], [21.26], atol = 0.1) # disk-based phase margin of 21.26 deg
```
This code never runs if slycot is not loaded => can you just remove it?

josiahdelange replied:
Yes - will fix.

murrayrm commented on control/tests/margin_test.py:
```diff
@@ -372,3 +373,135 @@ def test_stability_margins_discrete(cnum, cden, dt,
+    Lo = P*K # loop transfer function, broken at plant output
+    Li = K*P # loop transfer function, broken at plant input
+
+    if importlib.util.find_spec('slycot') == None:
```
You can just use `ct.slycot_check()` here and avoid using the `importlib` call.

murrayrm commented on control/margins.py:
```diff
@@ -517,3 +521,130 @@ def margin(*args):
 
     return margin[0], margin[1], margin[3], margin[4]
+
+def disk_margins(L, omega, skew = 0.0, returnall = False):
```
Remove spaces around `=` to be consistent with python-control (PEP8) coding conventions.

josiahdelange replied:
Will fix

murrayrm commented on control/margins.py:
```diff
@@ -517,3 +521,130 @@ def margin(*args):
+        to evaluate the disk-based stability margins
+    skew : float or array_like, optional
+        skew parameter(s) for disk margin calculation.
+        skew = 0 uses the "balanced" sensitivity function 0.5*(S - T)
+        skew = 1 uses the sensitivity function S
+        skew = -1 uses the complementary sensitivity function T
```
Which of these is the default value?

josiahdelange replied:
Will update.  Default of 0.0 is consistent with the MATLAB function `diskmargin`.

murrayrm commented on control/margins.py:
```diff
@@ -517,3 +521,130 @@ def margin(*args):
+        skew = 1 uses the sensitivity function S
+        skew = -1 uses the complementary sensitivity function T
+    returnall : bool, optional
+        If true, return frequency-dependent margins. If False (default),
```
true → True

josiahdelange replied:
Will fix

murrayrm commented on control/margins.py:
```diff
@@ -517,3 +521,130 @@ def margin(*args):
+    Example
+    --------
+    >> omega = np.logspace(-1, 3, 1001)
+    >> P = control.ss([[0,10],[-10,0]],np.eye(2),[[1,10],\
+    [-10,1]],[[0,0],[0,0]])
+    >> K = control.ss([],[],[],[[1,-2],[0,1]])
+    >> L = P*K
+    >> DM, DGM, DPM = control.disk_margins(L, omega, skew = 0.0)
```
Update to use PEP8 spacing: spaces after commas and operators (*), no space around = in keyword parameter assignments.  Also the `D` term in `P` can just be entered as `0` (it will be broadcast to the right size).

josiahdelange replied:
Will update.

murrayrm commented on control/margins.py:
```diff
@@ -517,3 +521,130 @@ def margin(*args):
+        raise ValueError("Loop gain must be square (n_inputs = n_outputs)")
+
+    # Need slycot if L is MIMO, for mu calculation
+    if (not L.issiso()) and (ab13md == None):
```
Parenthesis are redundant: `not L.issiso() and ab13md == None` will work.

josiahdelange replied:
Will fix

murrayrm commented on control/margins.py:
```diff
@@ -517,3 +521,130 @@ def margin(*args):
+
+    # Frequency-dependent complex disk margin, computed using upper bound of
+    # the structured singular value, a.k.a. "mu", of (S + (skew - I)/2).
+    DM = np.zeros(omega.shape, np.float64)
+    DGM = np.zeros(omega.shape, np.float64)
+    DPM = np.zeros(omega.shape, np.float64)
```
Use of `np.float64` not needed here?

josiahdelange replied:
Agree.  Will update

murrayrm commented on control/margins.py:
```diff
@@ -517,3 +521,130 @@ def margin(*args):
+    DM = np.zeros(omega.shape, np.float64)
+    DGM = np.zeros(omega.shape, np.float64)
+    DPM = np.zeros(omega.shape, np.float64)
+    for ii in range(0,len(omega)):
```
Missing space after comma (PEP8).

murrayrm commented on control/margins.py:
```diff
@@ -517,3 +521,130 @@ def margin(*args):
+    DPM = np.zeros(omega.shape, np.float64)
+    for ii in range(0,len(omega)):
+        # Disk margin (a.k.a. "alpha") vs. frequency
+        if L.issiso() and (ab13md == None):
```
Parentheses around second expression are not needed.

murrayrm commented on control/margins.py:
```diff
@@ -517,3 +521,130 @@ def margin(*args):
+                np.array(num_loops*[2]))[0]
+
+        # Disk-based gain margin (dB) and phase margin (deg)
+        with np.errstate(divide = 'ignore', invalid = 'ignore'):
```
Remove excess spaces around keyword assignment operator.

murrayrm commented on control/margins.py:
```diff
@@ -517,3 +521,130 @@ def margin(*args):
+
+    if returnall:
+        # Frequency-dependent disk margin, gain margin and phase margin
+        return (DM, DGM, DPM)
```
Parentheses not needed here.

murrayrm commented on control/margins.py:
```diff
@@ -517,3 +521,130 @@ def margin(*args):
+        if DPM.shape[0]:
+            pmidx = np.where(DPM == np.min(DPM))
+
+        return ((not DM.shape[0] and float('inf')) or np.amin(DM),
+            (not gmidx != -1 and float('inf')) or DGM[gmidx][0],
+            (not DPM.shape[0] and float('inf')) or DPM[pmidx][0])
```
I found this very hard to parse.  I wonder whether using if-else syntax might be simpler:
```
return (
    float('inf') if DM.shape[0] == 0 else np.amin(DM),
    float('inf') if gmidx == -1 else DGM[gmidx][0],
    float('inf') if DPM.shape[0] == 0 else DPM[pmidx][0])
```

josiahdelange replied:
Agree - thanks for the snippet, that is much cleaner.  I tried to emulate `stability_margins`, but had whittled it down somewhat.  Will update.

murrayrm reviewed (commented):
Some additional suggestions attached, mainly focused on coding style (to be consistent with the standard style in the package).

josiahdelange committed:
Address @murrayrm review comments.

josiahdelange committed:
Update formatting per PEP8/@murrayrm review comments.  Add additional reference on disk/ellipse-based margin calculations.

josiahdelange committed:
Follow-on to e8897f6fb57d1c9f7b7409055383083cdb59ae68: remove now-unnecessary import of importlib

josiahdelange commented:
@murrayrm (and @slivingston) thanks for taking time to review.  I _think_ everything's addressed - but let me know of anything else!

murrayrm reviewed (commented):
Found a few more style issues.

josiahdelange committed:
Update formatting per @murrayrm review comments

josiahdelange committed:
Merge branch 'python-control:main' into jdelange/disk-margins

josiahdelange committed:
Remove temporarily-added string from docstring

josiahdelange committed:
Minor tweak to docstring to fit the word 'function' back into the description of skew = 0.0

josiahdelange commented:
@murrayrm thanks and latest formatting issues are now fixed.  The one failing check "Slycot from source / build-linux (pull_request)" does not seem to be the disk margin tests.

murrayrm commented:
Agreed.  Tracking the failed unit test in #1161.  I'll go ahead and merge this since the fail check is unrelated.
